### PR TITLE
Add print_to_string() to Any{Value,Type}

### PR DESCRIPTION
**Pull request by seanyoung**
Make it easier to print a BasicValueEnum/BasicTypeEnum

### CODE REVIEW

**Review by TheDan64 — CHANGES_REQUESTED** [review]
This is a great idea; but I think it's the wrong approach. Instead, we should default impl `print_to_string` for the `Basic{Type,Value}` traits, then the types, values, and enums get them for free. And we should remove the existing individual methods on the type/values since they're superceeded by the trait.

`BasicType` [already impls](https://github.com/TheDan64/inkwell/blob/master/src/types/traits.rs#L124-L126) `ptr_type` btw, so `BasicTypeEnum` already gets it with a `BasicType` trait import. We should just remove the individual impls for that too.

Yay, deleting code!

**Comment by seanyoung** [comment]
Aha, I missed that I needed the `BasicType` trait in scope to use ptr_type() on a `BasicValueEnum`.

Anyway I reworked it. Let me know what you think.

**Review by TheDan64 — CHANGES_REQUESTED** [review]

File: src/types/traits.rs [hunk 1]
```diff
@@ -22,13 +23,17 @@ macro_rules! trait_type_set {
...
 /// Represents any LLVM type.
 pub trait AnyType<'ctx>: AsTypeRef + Debug {
     /// Returns an `AnyTypeEnum` that represents the current type.
     fn as_any_type_enum(&self) -> AnyTypeEnum<'ctx> {
         AnyTypeEnum::new(self.as_type_ref())
     }
+
+    /// Prints the definition of a Type to a `LLVMString`.
+    fn print_to_string(&self) -> LLVMString {
+        self.print_to_string()
```
[the remark this answers is not in the document]
Comment (seanyoung) [reply src/types/traits.rs]: Yes, that's obviously broken. I've fixed that and even tested it this time.

File: src/types/fn_type.rs [hunk 2]
```diff
@@ -9,7 +9,7 @@ use crate::AddressSpace;
 use crate::context::ContextRef;
 use crate::support::LLVMString;
 use crate::types::traits::AsTypeRef;
-use crate::types::{PointerType, Type, BasicTypeEnum};
+use crate::types::{AnyType, PointerType, Type, BasicTypeEnum};
```
Comment (TheDan64) [inline src/types/fn_type.rs:12]: Is this import necessary here?

File: src/values/array_value.rs [hunk 3]
```diff
@@ -6,7 +6,7 @@ use std::fmt;
 
 use crate::support::LLVMString;
 use crate::types::ArrayType;
-use crate::values::traits::AsValueRef;
+use crate::values::traits::{AnyValue, AsValueRef};
```
Comment (TheDan64) [inline src/values/array_value.rs:9]: Is this import necessary?

File: src/values/fn_value.rs [hunk 4]
```diff
@@ -20,7 +20,7 @@ use crate::basic_block::BasicBlock;
 use crate::debug_info::DISubprogram;
 use crate::module::Linkage;
 use crate::support::{to_c_str, LLVMString};
-use crate::types::{FunctionType, PointerType};
+use crate::types::{AnyType, FunctionType, PointerType};
```
Comment (TheDan64) [inline src/values/fn_value.rs:23]: 1. Is this import necessary?
2. Should get rid of `FunctionValue::print_to_string`

**Review by TheDan64 — CHANGES_REQUESTED** [review]
Last set of concerns - then we should be good to go!

Still needing removal:
* `FunctionValue::print_to_string`
* ~`PhiValue::print_to_string`~ jk you already did this one

(both of these impl `AnyValue`)

File: src/types/fn_type.rs (hunk 2 above)
Comment (seanyoung) [reply src/types/fn_type.rs:12]: Yes, `impl fmt::Debug` uses self.print_to_strng()

File: src/values/array_value.rs (hunk 3 above)
Comment (seanyoung) [reply src/values/array_value.rs:9]: Yes, `impl fmt::Debug` calls self.print_to_string()

File: src/values/fn_value.rs (hunk 4 above)
Comment (seanyoung) [reply src/values/fn_value.rs:23]: Yes, this import is necessary.

Note that `cargo clippy` in CI would tell you about unused imports.

**Comment by seanyoung** [comment]
I've removed phi_value/fn_type print_to_string()

**Review by TheDan64 — APPROVED** [review]
Looks good. Thanks! Will merge after CI finishes

**Comment by TheDan64** [comment]
@seanyoung looks like fn_value.rs needs AnyValue imported: https://travis-ci.org/github/TheDan64/inkwell/jobs/744760345

**Comment by seanyoung** [comment]
CI succeeded now

**Comment by TheDan64** [comment]
Thanks!

**Pull request merged**